### PR TITLE
Check for LazyLoader in Jinja exec-module-shorthand

### LINKED ISSUE #22444: LazyLoader breaks Jinja shorthand for referencing execution modules
Code:
```jinja
{% set is_true = salt.modules.test.ping() %}
```

Error:
```
Jinja variable 'salt.loader.LazyLoader object' has no attribute 'test'
```

### PR DESCRIPTION
**Pull request by whiteinge**
Also use hasattr, setattr instead of dictionary assignments.

Fixes #22444

### CODE REVIEW

**Comment by whiteinge** [comment]
Go go Jenkins!

**Comment by whiteinge** [comment]
Go Go Jenkins!

**Comment by whiteinge** [comment]
Go Jenkins Go!
Jenkins Go Go!
Go, go, Jenkins?
Goeth, Jenkins, with God.
Jenkins the gogo dancer!
GOG, 'o Jenkins, is a fine, DRM free video game vendor.
gogogogogogogjenkins

**Comment by thatch45** [comment]
Go Go Jenkins!

**Comment by whiteinge** [comment]
Jenkins ignoring Tom

**Comment by thatch45** [comment]
Go Go Jenkins!

**Comment by thatch45** [comment]
@whiteinge this keeps dying mid-tests, it might be having some adverse effects

**Comment by whiteinge** [comment]
This is clearly a Jenkins plot to embarrass me in front of my boss. I think it's safe to ignore.

Just rebased on the latest 2015.2. Let's see if that does anything.

The integration test I added runs locally. I can't seem to run the full test suite locally. (Does the suite even run on OS X?) The output from Jenkins to-date hasn't been the least bit helpful. Here's hoping this one is different.

**Comment by thatch45** [comment]
Still aborting early @whiteinge , I wonder what is up....

**Comment by thatch45** [comment]
Go Go Jenkins!

**Comment by thatch45** [comment]
the renderers don't really need to even use the lazy loader, if we switch to the static loader does this go away? loading the 10 or so renderers is not really a big deal

**Comment by whiteinge** [comment]
Removing the the actual change alleviates the abort. I can't begin to speculate why Jenkins is aborting or what that even means. My tiny change piggy-backs on other changes that I was not involved in (lazyloader, cmd.run -> cmd.script) and I don't have time to dive deeper into either of those until after next week. This is regression and should block 2015.2, IMO. Is there someone else who can take this on?

**Comment by whiteinge** [comment]
Change plus test restored in the above commit.

**Comment by thatch45** [comment]
but not using the lazy loader should fix it right? lets just do that

**Comment by whiteinge** [comment]
* The change works locally. The test works locally.
* I do not know why Jenkins is aborting prematurely.
* I do not know if switching to not use the lazy loader will fix it. That is reasonable speculation for a starting-point.
* I do not have additional time to devote to this until after _The Thing_ next week. If 2015.2 can wait until that time I will revisit, if not can this fix be reassigned?

Savvy?

**Comment by thatch45** [comment]
I will try changing it to the static loader since we can't afford to break the tests on jenkins, so I will make sure it is taken care of

**Comment by whiteinge** [comment]
Much appreciated, good sir.

**Comment by thatch45** [comment]
DOH! Of course, it has to be the functions, not the renderers.

File: salt/utils/templates.py [hunk 1]
```diff
@@ -59,9 +59,9 @@ def render_tmpl(tmplsrc,
         if 'salt' in kws:
             if 'cmd.run' in kws['salt'] and 'cmd.shell' in kws['salt']:
                 kws['salt']['cmd.run'] = kws['salt']['cmd.shell']
-            if 'run' in kws['salt'].get('cmd', {}) \
-                    and 'shell' in kws['salt'].get('cmd', {}):
-                kws['salt']['cmd']['run'] = kws['salt']['cmd']['shell']
+            if ('cmd' in kws['salt'] and hasattr(kws['salt']['cmd'], 'run')
```
Comment (thatch45) [inline salt/utils/templates.py]: This line will never resolve, which may be causing the tests to abort

Comment (whiteinge) [inline salt/utils/templates.py]: [See here](https://github.com/saltstack/salt/blob/ee9e618/salt/renderers/jinja.py#L300) for an explanation of the hasattr/setattr stuffs.

Comment (thatch45) [inline salt/utils/templates.py]: I mean it will never find cmd in kws['salt'], it should find kws['cmd.run'] should it not?

Comment (whiteinge) [inline salt/utils/templates.py]: It does that [here](https://github.com/saltstack/salt/pull/22447/files#diff-96d7010eb773dc57ad2c8eb89cd12807L60) but this line is for the short notation that was added in 2014.7.

File: salt/renderers/jinja.py [hunk 2]
```diff
@@ -293,7 +294,7 @@ def _split_module_dicts():
 
         {{ salt.cmd.run('uptime') }}
     '''
-    if not isinstance(__salt__, dict):
+    if not isinstance(__salt__, (dict, salt.loader.LazyLoader)):
         return __salt__
     mod_dict = dict(__salt__)
```
Comment (thatch45) [inline salt/renderers/jinja.py]: In my tests this line breaks the mod_dict, the lazy loader does not return a dict of funcs

File: salt/utils/templates.py (hunk 1 above)
Comment (thatch45) [inline salt/utils/templates.py]: I find myself confused...
I am trying more things...

File: salt/renderers/jinja.py (hunk 2 above)
Comment (whiteinge) [inline salt/renderers/jinja.py]: Odd. The isinstance check shouldn't break anything. You're right the
lazy loader does not return a dict of funcs, it returns a LazyLoader
class, thus the addition here.

**Comment by thatch45** [comment]
Go Go Jenkins!

**Comment by thatch45** [comment]
I think it is in ssh, are you running your local tests with --ssh ?

**Comment by thatch45** [comment]
The problem is a major performance degradation. Your loop is causing the lazy loader to reload all modules every time we call the jinja renderer, hence the timeout, every test is just taking longer to run. The solution is to add this directly to the lazy loader

**Comment by whiteinge** [comment]
Closing in deference to #22723.

**Pull request closed without merging**